ee/hogai/graph/entity_search/toolkit.py

```diff
@@ -0,0 +1,208 @@
+
+        result_dict = {
+            "name": extra_fields.get("name", f"{entity_type.upper()} {result_id}"),
+            "extra_fields": extra_fields,
+            "type": entity_type.title(),
+            "id": result_id,
+            "url": self.build_url(entity_type, result_id),
+        }
+
+        return yaml.dump(result_dict, default_flow_style=False, allow_unicode=True, sort_keys=False).strip()
+
+    async def _gather_bounded(self, limit: int, coros: list[object]):
+        sem = asyncio.Semaphore(limit)
+
+        async def run(coro):
+            async with sem:
+                return await coro
+
+        return await asyncio.gather(*(run(c) for c in coros), return_exceptions=True)
+
+    async def _search_single_entity(self, entity_type: str | EntityType, query: str) -> EntitySearchTaskResult:
+        entity_meta = ENTITY_MAP.get(entity_type)
+        if not entity_meta:
+            return EntitySearchTaskResult(
+                entity_type=entity_type,
+                results=[],
+                warning=f"Invalid entity type: {entity_type}. Will not search for this entity type.",
+            )
+
+        klass_qs, _ = await database_sync_to_async(class_queryset)(
+            view=self,
+            klass=entity_meta["klass"],
+            project_id=self._team.project_id,
+            query=query,
+            search_fields=entity_meta["search_fields"],
+            extra_fields=entity_meta["extra_fields"],
+        )
+
+        def evaluate_queryset(klass_qs=klass_qs):
+            return list(klass_qs[: self.MAX_ENTITY_RESULTS])
+
+        entity_results = await database_sync_to_async(evaluate_queryset)()
+
+        return EntitySearchTaskResult(entity_type=entity_type, results=entity_results, warning=None)
+
+    def _format_results_for_display(
+        self, query: str, entity_types: list[str | EntityType], results: list[dict], counts: dict[str, int]
+    ) -> str:
+        content = ""
+        if not results:
+            content += f"No entities found matching the query '{query}' for entity types {entity_types}"
+        else:
+            result_summary = []
+            for result in results:
+                result_summary.append(self._get_formatted_entity_result(result))
+
+            total_results = len(results)
+            content += FOUND_ENTITIES_MESSAGE_TEMPLATE.format(
+                total_results=total_results, entities_list="\n---\n".join(result_summary)
+            )
+
+            if counts:
+                content += ENTITY_TYPE_SUMMARY_TEMPLATE.format(
+                    entity_type_summary="\n".join(
+                        [f"- {entity_type.title()}: {count}" for entity_type, count in counts.items() if count > 0]
+                    )
+                )
+            content += f"\n\n{HYPERLINK_USAGE_INSTRUCTIONS}"
+        return content
+
+    async def search_entities(self, query: str, entity_types: list[str | EntityType]) -> str:
```

Comment:
Let's also pivot here a bit. It shouldn't take a list of search queries, but just the query and entity type. Parallel tools will effectively handle both synchronous and asynchronous queries so that the root node will handle the parallelization (coming this week).

```suggestion
    async def execute(self, entity: EntityType, query: str) -> str:
```